src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,296 @@
+        """
+        Puts transactions in list, calls required function in case of range of blocks.
+        """
+        settlement_hashes_list = []
+        if settlement_hash is not None:
+            settlement_hashes_list = [settlement_hash]
+        elif start_block is not None and end_block is not None:
+            settlement_hashes_list = self.get_hashes_by_block(start_block, end_block)
+            # At this point we have all the needed hashes
+        self.decoder(settlement_hashes_list)
+
+    def decoder(self, settlement_hashes_list):
+        for settlement_hash in settlement_hashes_list:
+            self.decode_single_hash(settlement_hash)
+        return
+
+    def get_order_data_by_hash(self, settlement_hash):
+        """
+        Returns competition endpoint data since we need order_id and auction_id, and also AWS
+        Bucket response.
+        """
+        bucket_response = None
+        comp_data = None
+        try:
+            comp_data = requests.get(
+                f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+            )
+            status_code = comp_data.status_code
+            if status_code == 200:
+                comp_data = comp_data.json()
+                auction_id = comp_data["auctionId"]
+                bucket_response = dict(
+                    requests.get(
+                        f"https://solver-instances.s3.eu-central-1.amazonaws.com/prod-mainnet/{auction_id}.json"
+                    ).json()
+                )
+            elif status_code == 404:
+                comp_data = requests.get(
+                    f"https://barn.api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+                )
+                status_code = comp_data.status_code
+                if comp_data.status_code == 200:
+                    comp_data = comp_data.json()
+                    auction_id = comp_data["auctionId"]
+                    bucket_response = dict(
+                        requests.get(
+                            f"https://solver-instances.s3.eu-central-1.amazonaws.com/staging-mainnet/{auction_id}.json"
+                        ).json()
+                    )
+            return comp_data["solutions"][-1]["orders"], bucket_response
+        except ValueError as except_err:
+            self.logger.error("Unhandled exception: %s", str(except_err))
+
+    def extract_win_trade_data(self, decoded_settlement, trade, order_type):
+        sell_token_index = trade["sellTokenIndex"]
+        buy_token_index = trade["buyTokenIndex"]
+        sell_token_clearing_price = decoded_settlement.clearing_prices[sell_token_index]
+        buy_token_clearing_price = decoded_settlement.clearing_prices[buy_token_index]
+        return self.get_surplus(
+            trade, sell_token_clearing_price, buy_token_clearing_price, order_type
+        )
+
+    def extract_qmdo_trade_data(self, trade, order, instance2, order_type):
```

Comment:
also split this function.
naming: `quasimodo` instead of `qmdo`. or even better: make it solver agnostic. all external solvers could be used for this test, right?

Reply:
`instance2` should be `instance`